### PR TITLE
Reading/Editing/Octaning “Testing” section

### PR DESCRIPTION
**Pull request by muziejus**
Working on #588

### CODE REVIEW

**Comment by muziejus** [comment]
I don't know how to properly separate two PRs at once, so the first five commits here come from my other current PR. Nevertheless, I'm using this space to take some notes on the testing pages:

1. In the "Introduction," I think that `this.owner` is described too abruptly.

> In the example below, we import the setupTest function from ember-qunit and call it with the hooks object to setup the test context with access to the this.owner property. This provides us direct container access to interact with Ember's Dependency Injection system.

The link to the Dependency Injection guide is handy, but it's not quite clear why, if `this.owner` is a property of the test, that it "owns" an `ApplicationInstance`. It may be better to use text like:

> In the example below, we import the setupTest function from ember-qunit and call it with the hooks object to setup the test context. This uses Ember's Dependency Injection system to create an ApplicationInstance that serves as the "owner" for objects that are instantiated during testing. We can access the ApplicationInstance with the `this.owner` property to "look up" (i.e. instantiate and / or retrieve) objects.

**Comment by muziejus** [comment]
Next, the testing sections present a few difficulties for octanization:

2. In "Testing Basics," there is a whole section to testing observers. As @pzuraq notes, "[observers and event listeners are not a recommended pattern anymore](https://www.pzuraq.com/coming-soon-in-ember-octane-part-1-native-classes/)," so maybe this section should be cut entirely or refactored to use `@tracked`.

3. The first test in "Testing Components" relies on `attributeBindings`. These are, if I understand correctly, not to be used in Octane, as something like a component's `style` attribute can be set programmatically with something like `<div style={{this.style}}>…</div>`. I would propose a simpler test that uses Glimmer component basics.

4. I don't know what the Octane orthodoxy is vis-à-vis controllers, but given that the Controllers section of the guides is not on the sidebar, that it has not been updated to use native class syntax, and that the controllers section of the "Routing" guide is gone, I think perhaps it's time to retire the controllers testing section, as well?

File: guides/release/testing/testing-components.md [hunk 1]
```diff
@@ -156,23 +156,23 @@ Imagine you have the following component that changes its title when a button is
...
+    this.set('title', 'This is Magic');
+  };
+};
 ```
 
 ```handlebars {data-filename="app/templates/components/magic-title.hbs"}
 <h2>{{this.title}}</h2>
 
-<button class="title-button" {{action "updateTitle"}}>
+<button class="title-button" onclick={{action this.updateTitle}}>
```
Comment (rwjblue) [inline guides/release/testing/testing-components.md]: We are avoiding `onclick=` usage (see @mixonic's awesome explanation over at https://github.com/ember-learn/guides-source/pull/598#discussion_r265390517)

```suggestion
<button class="title-button" {{action this.updateTitle}}>
```

File: guides/release/testing/testing-components.md [hunk 2]
```diff
@@ -218,21 +218,21 @@ passing along the form's data:
...
+  comment = '';
 
-  actions: {
-    submitComment() {
-      this.submitComment({ comment: this.comment });;
-    }
-  }
+  @action
+  submitComment() {
+    this.submitComment({ comment: this.comment });;
```
Comment (rwjblue) [inline guides/release/testing/testing-components.md]: On glimmer components, the passed in arguments are not on `this`, they are in `this.args`.

```suggestion
    this.args.submitComment({ comment: this.comment });;
```

File: guides/release/testing/testing-components.md [hunk 3]
```diff
@@ -218,21 +218,21 @@ passing along the form's data:
...
+  };
```
Comment (rwjblue) [inline guides/release/testing/testing-components.md]: Inside a class body there is no semicolon or comma separating methods.

```suggestion
  }
```

File: guides/release/testing/testing-components.md [hunk 4]
```diff
@@ -218,21 +218,21 @@ passing along the form's data:
...
 });
 ```
 
 ```handlebars {data-filename="app/templates/components/comment-form.hbs"}
-<form {{action "submitComment" on="submit"}}>
+<form onsubmit={{action this.submitComment}}>
```
Comment (rwjblue) [inline guides/release/testing/testing-components.md]:
```suggestion
<form {{action this.submitComment on="submit"}}>
```

File: guides/release/testing/testing-components.md [hunk 5]
```diff
@@ -285,22 +285,24 @@ Imagine you have the following component that uses a location service to display
...
+export default class LocationIndicatorComponent extends Component {
+  @service locationService;
 
   // when the coordinates change, call the location service to get the current city and country
-  city: computed('locationService.currentLocation', function () {
+  @computed('locationService.currentLocation')
+  get city() {
     return this.locationService.getCurrentCity();
-  }),
+  });
```
Comment (rwjblue) [inline guides/release/testing/testing-components.md]:
```suggestion
  }
```

File: guides/release/testing/testing-components.md [hunk 6]
```diff
@@ -285,22 +285,24 @@ Imagine you have the following component that uses a location service to display
...
 
-  country: computed('locationService.currentLocation', function () {
+  @computed('locationService.currentLocation')
+  get country() {
     return this.locationService.getCurrentCountry();
-  })
-});
+  });
```
Comment (rwjblue) [inline guides/release/testing/testing-components.md]:
```suggestion
  }
```

File: guides/release/testing/testing-components.md [hunk 7]
```diff
@@ -465,17 +467,17 @@ Imagine you have a typeahead component that uses [`Ember.run.debounce`](https://
...
-      debounce(this, this.fetchResults, this.searchValue, 250);
-    }
-  }
-});
+export default class DelayedTypeaheadComponent extends Component {
+  @action 
+  handleTyping() {
+    //the fetchResults function is passed into the component from its parent
+    debounce(this, this.fetchResults, this.searchValue, 250);
+  };
```
Comment (rwjblue) [inline guides/release/testing/testing-components.md]:
```suggestion
  }
```

File: guides/release/testing/unit-testing-basics.md [hunk 8]
```diff
@@ -17,13 +17,14 @@ based on a `foo` property.
...
-  foo: 'bar',
+export default class SomeThingService extends Service {
+  foo = 'bar';
 
-  computedFoo: computed('foo', function() {
-    return `computed ${this.foo}`;
+  @computed('foo')
+  get computedFoo() {
+    return `${this.foo}`;
   })
```
Comment (rwjblue) [inline guides/release/testing/unit-testing-basics.md]:
```suggestion
  }
```

File: guides/release/testing/unit-testing-basics.md [hunk 9]
```diff
@@ -138,14 +139,15 @@ Suppose we have an object that has a property and a method observing that proper
 import Service from '@ember/service';
 import { observer } from "@ember/object";
 
-export default Service.extend({
-  foo: 'bar',
-  other: 'no',
+export default class SomeThingService extends Service {
```
Comment (rwjblue) [inline guides/release/testing/unit-testing-basics.md]: Lets revert this change for now, I don't think we can actually talk about observers in the new world (because there is no decorator form of `observes` in core, and ultimately folks should _probably_ move away from them).

**Comment by rwjblue** [comment]
> In "Testing Basics," there is a whole section to testing observers. As @pzuraq notes, "observers and event listeners are not a recommended pattern anymore," so maybe this section should be cut entirely or refactored to use @tracked.

Ya, agreed. Lets remove the observer test example section.

**Comment by rwjblue** [comment]
> The first test in "Testing Components" relies on attributeBindings. These are, if I understand correctly, not to be used in Octane, as something like a component's style attribute can be set programmatically with something like <div style={{this.style}}>…</div>. I would propose a simpler test that uses Glimmer component basics.

I actually think the example there is still valid. In a `@glimmer/component` world `style` attribute does not have any special merging behaviors. If you wanted to have a component that took `color` as an argument, it would still _mostly_ need the same code.

I'd write the example component as:

```js
import Component from '@glimmer/component';

export default class PrettyColor extends Component {
  get style() {
    return `color: ${this.args.name}`;
  }
});
```

```hbs
<div style={{this.style}}>
  Pretty Color: {{@name}}
</div>
```

**Comment by rwjblue** [comment]
> but given that the Controllers section of the guides is not on the sidebar, that it has not been updated to use native class syntax

I think it still needs to be updated, octane does not remove the need for controllers at all. However, since the controllers section in the sidebar is removed, perhaps we should just remove the controller testing section from the sidebar for now as well. That way we can bring both back together, when they are ready.

File: guides/release/testing/unit-testing-basics.md [hunk 10]
```diff
@@ -65,13 +66,13 @@ the `foo` property).
...
-export default Service.extend({
-  foo: 'bar',
+export default class SomeThingService extends Service {
+  foo = 'bar';
 
   testMethod() {
     this.set('foo', 'baz');
   }
-});
+};
```
Comment (efx) [inline guides/release/testing/unit-testing-basics.md]:
```suggestion
}
```

File: guides/release/services/services.md [hunk 11]
```diff
@@ -25,8 +25,8 @@ Services must extend the [`Service`](https://www.emberjs.com/api/ember/release/m
 ```javascript {data-filename=app/services/shopping-cart.js}
 import Service from '@ember/service';
 
-export default Service.extend({
-});
+export default class ShoppingCartService extends Service {
+};
```
Comment (efx) [inline guides/release/services/services.md]:
```suggestion
}
```

**Comment by muziejus** [comment]
@rwjblue and @efx, any thoughts on adding a bit more explanation to `this.owner` in the introduction section? The controllers test also describes it better than the intro, imo:

> Next we use the owner API to gain access to the controller we'd like to test.
Using the `this.owner.lookup` method we get the instance of the `PostsController` and can check the action in our test.
The `this.owner.lookup` helper returns objects generated by the framework in your applications
and is also exposed in tests for your usage. Here it will return a singleton instance of the `PostsController`.

**Comment by rwjblue** [comment]
OK, I think we need a rebase now (to drop out the commits that were also in #588). Does that make sense? Also, happy to help with that if you'd like...

**Comment by muziejus** [comment]
@rwjblue I think I got it.

**Comment by muziejus** [comment]
Thanks, all! 🎸

**Comment by efx** [comment]
@muziejus I think discussing how we introduce the owner API is a good idea. I had a similar problem with how the current guides talked about services: https://github.com/ember-learn/guides-source/issues/300#issue-386970151
I prefer linking to the better definition and even the API documentation itself. Having 2 definitions of an API floating around in the guides could lead to confusion as you rightly note.

**Comment by muziejus** [comment]
@efx Good point about not duplicating definitions, as I sort of do that in my proposal, copy and pasting from the dependency injection page's discussion of `ApplicationInstance`.

Could the glossary help here? 

Similarly, something like:

```markdown
In the example below, we import the `setupTest` function from `ember-qunit` and call it with the `hooks` object to setup the test context. This provides us with the `this.owner` property. As the [API docs note]:

> When Ember's [Dependency Injection] system creates an [ApplicationInstance], it exposes a `this.owner` property that serves as the "owner" for objects that are instantiated during testing. We can access the `ApplicationInstance` with the `this.owner.lookup` to "look up" (i.e. instantiate and / or retrieve) objects. Similarly, we can register...
```

Might be good, since it marks that we're relying on a different document for the quoted material. That may be friendlier than just a link.

File: guides/release/testing/unit-testing-basics.md [hunk 12]
```diff
@@ -65,13 +66,13 @@ the `foo` property).
 ```javascript {data-filename=app/services/some-thing.js}
 import Service from '@ember/service';
```
Comment (rwjblue) [inline guides/release/testing/unit-testing-basics.md]:
```suggestion
import Service from '@ember/service';
import { tracked } from '@glimmer/tracking';
```

File: guides/release/testing/unit-testing-basics.md [hunk 13]
```diff
@@ -65,13 +66,13 @@ the `foo` property).
...
 
-export default Service.extend({
-  foo: 'bar',
+export default class SomeThingService extends Service {
+  foo = 'bar';
```
Comment (rwjblue) [inline guides/release/testing/unit-testing-basics.md]: This should be a tracked property:

```suggestion
  @tracked foo = 'bar';
```

File: guides/release/testing/unit-testing-basics.md [hunk 14]
```diff
@@ -65,13 +66,13 @@ the `foo` property).
...
 
   testMethod() {
     this.set('foo', 'baz');
```
Comment (rwjblue) [inline guides/release/testing/unit-testing-basics.md]:
```suggestion
    this.foo = 'baz';
```

File: guides/release/testing/unit-testing-basics.md [hunk 15]
```diff
@@ -102,14 +103,14 @@ that returns a value based on some internal state.
 ```javascript {data-filename=app/services/some-thing.js}
 import Service from '@ember/service';
 
-export default Service.extend({
-  count: 0,
+export default class SomeThingService extends Service {
+  count = 0;
```
Comment (rwjblue) [inline guides/release/testing/unit-testing-basics.md]:
```suggestion
  @tracked count = 0;
```

File: guides/release/testing/unit-testing-basics.md [hunk 16]
```diff
@@ -102,14 +103,14 @@ that returns a value based on some internal state.
 ```javascript {data-filename=app/services/some-thing.js}
 import Service from '@ember/service';
```
Comment (rwjblue) [inline guides/release/testing/unit-testing-basics.md]:
```suggestion
import Service from '@ember/service';
import { tracked } from '@glimmer/tracking';
```

File: guides/release/testing/unit-testing-basics.md [hunk 17]
```diff
@@ -102,14 +103,14 @@ that returns a value based on some internal state.
...
 
-export default Service.extend({
-  count: 0,
+export default class SomeThingService extends Service {
+  count = 0;
 
   calc() {
     this.incrementProperty('count');
```
Comment (rwjblue) [inline guides/release/testing/unit-testing-basics.md]:
```suggestion
    this.count += 1;
```

**Comment by efx** [comment]
@muziejus big 👍 from me to your revision. Honestly don't think we even need to quote the API documentation as it is linked. I also want to say we should continue any wording tweaks outside of this PR because the *grand work* you're doing here should be be merged once the technical review notes are done.

File: guides/release/testing/testing-components.md [hunk 18]
```diff
@@ -8,20 +8,19 @@ The `style` attribute of the component is bound to its `style` property.
...
 
-  style: computed('name', function() {
-    return `color: ${this.name}`;
-  })
-});
+export default class PrettyColorComponent extends Component {
+  get style() {
+    return `color: ${this.args.name}`;
+  }
+};
```
Comment (efx) [inline guides/release/testing/testing-components.md]:
```suggestion
}
```

File: guides/release/testing/testing-components.md [hunk 19]
```diff
@@ -156,23 +155,23 @@ Imagine you have the following component that changes its title when a button is
...
-    updateTitle() {
-      this.set('title', 'This is Magic');
-    }
-  }
-});
+  @action
+  updateTitle() {
+    this.set('title', 'This is Magic');
+  };
+};
```
Comment (efx) [inline guides/release/testing/testing-components.md]:
```suggestion
}
```

File: guides/release/testing/testing-components.md [hunk 20]
```diff
@@ -156,23 +155,23 @@ Imagine you have the following component that changes its title when a button is
...
-  actions: {
-    updateTitle() {
-      this.set('title', 'This is Magic');
-    }
-  }
-});
+  @action
+  updateTitle() {
+    this.set('title', 'This is Magic');
+  };
```
Comment (efx) [inline guides/release/testing/testing-components.md]:
```suggestion
  }
```
I think method definitions and the class exports do not require a semicolon so we should remove them. You can see the bulk of the examples in https://github.com/ember-learn/guides-source/blob/octane/guides/release/upgrading/editions.md#native-classes too for reference.

File: guides/release/testing/testing-components.md [hunk 21]
```diff
@@ -218,21 +217,21 @@ passing along the form's data:
...
 
-  actions: {
-    submitComment() {
-      this.submitComment({ comment: this.comment });;
-    }
+  @action
+  submitComment() {
+    this.args.submitComment({ comment: this.comment });;
   }
 });
```
Comment (efx) [inline guides/release/testing/testing-components.md]:
```suggestion
}
```

File: guides/release/testing/testing-components.md (hunk 20 above)
Comment (muziejus) [reply guides/release/testing/testing-components.md]: Right. I have been remiss in making sure I do this everywhere.

**Comment by muziejus** [comment]
@rwjblue Per your rec, I deleted (4e8badb) the section on testing observers, but I do think it'd be useful to instead replace it with a section on testing tracked properties, so that may be a way to advance the points you bring up.

**Comment by muziejus** [comment]
@efx Thanks. You're right… I think once the code is somewhat stable I can have a stab at the language in separate PRs to work on better comprehension. As a relative noob, I think I have a couple insights into sniffing out where things get too confusing too quickly!
